x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/TopValuesList.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    private Expression limitField() {
+        return parameters().get(0);
+    }
+
+    private Expression orderField() {
+        return parameters().get(1);
+    }
+
+    private int limitValue() {
+        return (int) limitField().fold();
+    }
+
+    private String orderRawValue() {
+        return BytesRefs.toString(orderField().fold());
+    }
+
+    private boolean orderValue() {
+        return orderRawValue().equalsIgnoreCase(ORDER_ASC);
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        var typeResolution = isType(
+            field(),
+            dt -> dt == DataType.DATETIME || dt.isNumeric() && dt != DataType.UNSIGNED_LONG,
+            sourceText(),
+            FIRST,
+            "numeric except unsigned_long or counter types"
+        ).and(isFoldable(limitField(), sourceText(), SECOND))
+            .and(isType(limitField(), dt -> dt == DataType.INTEGER, sourceText(), SECOND, "integer"))
+            .and(isFoldable(orderField(), sourceText(), THIRD))
+            .and(isString(orderField(), sourceText(), THIRD));
+
+        if (typeResolution.unresolved()) {
+            return typeResolution;
+        }
+
+        var limit = limitValue();
+        var order = orderRawValue();
+
+        if (limit <= 0) {
+            return new TypeResolution(format(null, "Limit must be greater than 0. Got {}", limit));
+        }
+
+        if (order.equalsIgnoreCase(ORDER_ASC) == false && order.equalsIgnoreCase(ORDER_DESC) == false) {
+            return new TypeResolution(format(null, "Invalid order value. Expected [{}, {}] but got {}", ORDER_ASC, ORDER_DESC, order));
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public DataType dataType() {
+        return field().dataType();
+    }
+
+    @Override
+    protected NodeInfo<TopValuesList> info() {
+        return NodeInfo.create(this, TopValuesList::new, children().get(0), children().get(1), children().get(2));
+    }
+
+    @Override
+    public TopValuesList replaceChildren(List<Expression> newChildren) {
+        return new TopValuesList(source(), newChildren.get(0), newChildren.get(1), newChildren.get(2));
+    }
+
+    @Override
+    public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+        DataType type = field().dataType();
+        if (type == DataType.LONG || type == DataType.DATETIME) {
+            return new TopValuesListLongAggregatorFunctionSupplier(inputChannels, limitValue(), orderValue());
+        }
+        if (type == DataType.INTEGER) {
+            return new TopValuesListIntAggregatorFunctionSupplier(inputChannels, limitValue(), orderValue());
+        }
+        if (type == DataType.DOUBLE) {
+            return new TopValuesListDoubleAggregatorFunctionSupplier(inputChannels, limitValue(), orderValue());
+        }
```

Comment:
I'll bet folks will want `BytesRef` one day. but it can wait.

Reply:
As I'm surrogating to Max/Min, I went the path of having TopValues with the same types ad Max/Min, and later we could add the extras to everything. Not that one really depends on the other, but for convenience and keeping the PR "short"

Reply:
Created this issue: https://github.com/elastic/elasticsearch/issues/109849